src/pint/models/model_builder.py

```diff
@@ -26,276 +26,296 @@ class UnknownBinaryModel(ValueError):
+                unrec_param.append(pp)
+
+        param_components_inpar = {p: self.param_component_map[p] for p in pint_par}
+        # Back map the possible_components and the parameters in the parfile
+        # This will remove the duplicate components.
+        possible_components = {}
+        for k, cps in param_components_inpar.items():
+            for cp in cps:
+                if cp == 'timing_model': # skip the timing model parameters.
+                    continue
+                if cp in possible_components.keys():
+                    possible_components[cp].append(k)
                 else:
-                    # Check if the prefix match
-                    for p in self.control_params.keys():
-                        if p.startswith(ctrlp):
-                            info[ctrlp] = self.param_unrecognized[ctrlp]
-        return info
+                    possible_components[cp] = [k]
+        # Get conflicting components based on categories
+        # conflict_components = {category: [component1, component2, ...]}
+        conflict_components = {}
+        for k in possible_components.keys():
+            cate = self.component_category_map[k]
+            if cate in conflict_components.keys():
+                conflict_components[cate].append(k)
+            else:
+                conflict_components[cate] = [k]
+        # 3. resolve the conflicting components, mulitple components in the same
+        # category. They generally share some same parameters
+        resolved = []
+        for ca, cps in conflict_components.items():
+            # skip the pulsar system, since it is selected by the BINARY parameter
+            if ca == 'pulsar_system':
+                if binary is not None:
+                    raise ValueError("Pulsar system is set by parameter `BINARY`.")
+                resolved.append(ca)
+                continue
+            # Only one component selected, no confilicte
+            if len(cps) == 1:
+                result_componets.append(self.components[cps[0]])
+                resolved.append(ca)
+                continue
+            # There is a conflict in the categories
+            rank = self._get_param_overlap_rank(cps, pint_par, 'component_special_params')
+            # First check the special parameters, the most special-parameter
+            #-overlap component will be selected
+            # Check if there are more than one components have the same rank
+            # If the highest rank has the same rank with the second highest,
+            # the conflict is not resolved by speical parameters.
+            if rank[1][1] != rank[0][1]: # has an unique highest case.
+                result_componets.append(self.components[rank[0][0]])
+                resolved.append(ca)
+            else:
+                # First no overlap case, which should mean no special parameters
+                # in the component, will do a search depends on the overlap of
+                # all parameters.
```

Comment:
I think - if I understand this correctly - this is counting how many parameters overlap between what's in the par file and what's in the component? This is not an easy rule for users to understand. I suggest raising an exception for any ambiguous cases where a distinctive parameter doesn't force the choice. (e.g. NE_SW is only in solar_wind_dispersion so it is required.)

Reply:
Yes, Counting the overlap is the idea. I am very hesitant to add any case-specific checks here in this general code. We can have a separate function to do the job of hard fail.

Reply:
I don't think counting overlap is a good idea at all. I think it will surprise users when their model suddenly switches because they added a parameter that pushes them over the threshold. My idea was that for every model there should be at least one parameter not present in any alternatives, and that these parameters should force the selection. (For example, if you have NE_SW, you clearly need a solar wind dispersion model. No special casing here because that parameter only occurs in that model.) If you have two alternatives for which one has a subset of the parameters of the other, the one with the smaller parameter set should be chosen unless there is a parameter from the larger. Are there any cases where models cannot be distinguished in this way?

Reply:
The component_special_parameter is meant to tell the model builder that choose one component vs the other one. The current setup is that it will check the over_lap of component_special_parameters first, if the conflict still exists, it will check the overlap of all parameters. I can remove the latter check and raise a conflict error.

Reply:
I did update the code here. I think it is much simpler.